### PR TITLE
chainntfns: add new option for conf notifications to send block 

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we add a new option for the existing confirmation
notification system that optionally allows the caller to specify that a
block should be included as well.

The only quirk w/ the implementation here is the neutrino backend:
usually we get filtered blocks, we so need to first fetch the block
again so we can deliver the full block to the notifier. On the notifier
end, it'll only be checking for the transactions we care about, to
sending a full block doesn't affect the correctness.

We also extend the `testBatchConfirmationNotification` test to assert
that a block is only included if the caller specifies it.

In this commit, we add a test case to exercise the fact that if we
request a confirmation notification with the raw block, then it includes
one. We add this to the existing test that uses the conf ntfn rpc call
for simplicity.

### CODE REVIEW

File: chainntnfs/interface.go [hunk 1]
```diff
@@ -67,6 +67,32 @@ func (t TxConfStatus) String() string {
 	}
 }
 
+// notifierOptions is a set of functional options that allow callers to further
+// modify the type of chain even notifications they receive.
```
Comment (guggero) [inline chainntnfs/interface.go:71]: nit: s/even/event/

File: chainntnfs/interface.go [hunk 2]
```diff
@@ -67,6 +67,32 @@ func (t TxConfStatus) String() string {
...
+type notifierOptions struct {
+	// includeBlock if true, then the dispatched confirmation notification
+	// will include the block that mined the transaction.  includeBlock
```
Comment (guggero) [inline chainntnfs/interface.go:74]: nit: looks like some leftover sentence fragment.

File: chainntnfs/neutrinonotify/neutrino.go [hunk 3]
```diff
@@ -649,11 +649,19 @@ func (n *NeutrinoNotifier) historicalConfDetails(confRequest chainntnfs.ConfRequ
...
-	// block, which will handle all of the notification logic for us.
-	err := n.txNotifier.ConnectTip(
-		&newBlock.hash, newBlock.height, newBlock.txns,
-	)
+	// We'll extend the txNotifier's height with the information of this
+	// new block, which will handle all of the notification logic for us.
+	//
+	// We actually need the _full_ block here as well in order to be able
+	// to send the full block back up to the client. The neutrino client
+	// itself will only dispathc  a block if one of the items we're looking
```
Comment (guggero) [inline chainntnfs/neutrinonotify/neutrino.go:657]: nit: `s/dispathc  a/dispatch a/`

File: chainntnfs/txnotifier.go [hunk 4]
```diff
@@ -1548,13 +1583,14 @@ func (n *TxNotifier) filterTx(tx *btcutil.Tx, blockHash *chainhash.Hash,
...
-				blockHeight, blockHash)
+				blockHeight, block.Hash())
 
 			details := &TxConfirmation{
 				Tx:          tx.MsgTx(),
-				BlockHash:   blockHash,
+				BlockHash:   block.Hash(),
 				BlockHeight: blockHeight,
 				TxIndex:     uint32(tx.Index()),
+				Block:       block.MsgBlock(),
```
Comment (guggero) [inline chainntnfs/txnotifier.go:1593]: I think we might also need to set this in `interface.go/ConfDetailsFromTxIndex()`, `bitcoind.go/confDetailsManually()`, `btcd.go/confDetailsManually()` and `neutrino.go/historicalConfDetails()`, unless I'm misunderstanding something.

File: lntest/itest/lnd_taproot_test.go [hunk 5]
```diff
@@ -1297,25 +1297,31 @@ func confirmAddress(ctx context.Context, t *harnessTest,
...
 
-	// We now expect our confirmation to go through.
+	// We now expect our confirmation to go through, and also that the
+	// block was specified.
 	confMsg, err := confClient.Recv()
 	require.NoError(t.t, err)
 	conf := confMsg.GetConf()
 	require.NotNil(t.t, conf)
 	require.Equal(t.t, conf.BlockHeight, uint32(currentHeight+1))
+	require.NotNil(t.t, conf.RawBlock)
```
Comment (guggero) [inline lntest/itest/lnd_taproot_test.go:1324]: Should we make sure the block can be decoded and has the expected height?

**Review by guggero — COMMENTED** [review]
Nice and compact diff.

The unit test seem to fail because a mock wasn't updated.

**Comment by guggero** [comment]
Should be rebased to get latest updates from master branch.

File: lntest/itest/lnd_taproot_test.go (hunk 5 above)
Comment (Roasbeef) [reply lntest/itest/lnd_taproot_test.go:1324]: Added the block, decode, but which height?

File: chainntnfs/txnotifier.go (hunk 4 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:1593]: Nice catch! Just fixed those areas, will also tighten up the unit tests to ensure they're caught.

**Comment by Roasbeef** [comment]
Rebased!

File: lntest/itest/lnd_taproot_test.go (hunk 5 above)
Comment (guggero) [reply lntest/itest/lnd_taproot_test.go:1324]: Ah, I always forget that the block height isn't a property of the block itself :see_no_evil:

**Review by guggero — APPROVED** [review]
LGTM :tada: 
Needs an entry in the change log, otherwise this is ready.

File: chainntnfs/interface.go [hunk 6]
```diff
@@ -700,26 +733,38 @@ func ConfDetailsFromTxIndex(chainConn TxIndexConn, r ConfRequest,
...
 		return nil, TxNotFoundIndex,
 			fmt.Errorf("unable to get block with hash %v for "+
 				"historical dispatch: %v", blockHash, err)
 	}
 
+	// In the modern chain (the only one we really care about for LN), the
+	// coinbase transaction of all blocks will include the block height.
+	// Therefore we can save another query, and just use that height
+	// directly.
+	blockHeight, err := blockchain.ExtractCoinbaseHeight(
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply chainntnfs/interface.go:747]: BIP 34 saves the day again! Same thing is used in Taro to be able to generate stabe/unique asset IDs

**Comment by Roasbeef** [comment]
Release notes added!

File: chainntnfs/interface.go [hunk 7]
```diff
@@ -67,6 +69,31 @@ func (t TxConfStatus) String() string {
...
+// defaultNotifierOptions returns the set of default options for the notifier.
+func defaultNotifierOptions() *notifierOptions {
+	return &notifierOptions{}
+}
+
+// NotifierOption is a functional option that allows a caller to modify the
+// events received from the notifier.
+type NotifierOption func(*notifierOptions)
+
+// WithIncludeBlock is an optional argument that allows the calelr to specify
```
Comment (positiveblue) [inline chainntnfs/interface.go:89]: nit: s/calelr/caller/

File: chainntnfs/txnotifier.go [hunk 8]
```diff
@@ -622,16 +628,22 @@ func (n *TxNotifier) newConfNtfn(txid *chainhash.Hash,
...
+	optFuncs ...NotifierOption) (*ConfRegistration, error) {
 
 	select {
 	case <-n.quit:
 		return nil, ErrTxNotifierExiting
 	default:
 	}
 
+	opts := defaultNotifierOptions()
+	for _, optFunc := range optFuncs {
```
Comment (positiveblue) [inline chainntnfs/txnotifier.go:641]: This pattern is really nice :+1:

File: chainntnfs/txnotifier.go [hunk 9]
```diff
@@ -1548,13 +1582,14 @@ func (n *TxNotifier) filterTx(tx *btcutil.Tx, blockHash *chainhash.Hash,
 		notifyDetails := func(confRequest ConfRequest) {
 			Log.Debugf("Found initial confirmation of %v: "+
 				"height=%d, hash=%v", confRequest,
-				blockHeight, blockHash)
+				blockHeight, block.Hash())
```
Comment (positiveblue) [inline chainntnfs/txnotifier.go:1585]: nit: maybe it's worth commenting in the godoc that `block` cannot be nil if `onConf != nil`?

**Review by positiveblue — APPROVED** [review]
LGTM :rocket: nothing blocking

**Pull request merged**